### ISSUE TITLE
find a suitable geocode for ALC's

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#10387]
>But let  me explain what my struggle was with the subject of "geocode" all the time. Here is one example. There are many more like this:
>
> ```
>main/src/cgeo/geocaching/CacheDetailActivity.java:1173
>
>addContextMenu(details.add(R.string.cache_geocode, cache.getGeocode()).right);
>```
>
> `addContextMenu` can't be just modified like this:
>
> `addContextMenu(details.add(R.string.cache_geocode, cache.getShortGeocode()).right)`
>
> as it needs the **full** geocode to do its work. Since it does not get a reference to the `cache` object, it cannot do a `cache.getShortGeocode()` at the place where it is needed to bind the string to the ui property.
>
> Of course I could have added another parameter to addContextMenu and pass the `cache.getShortGeocode()` in addition but that can't really be the solution or can it?

Ok, it really seems to be not that easy as I would have expected it. Need to think about it...

> I remember there was a reference code in the beginning, which the user could even define hisself.
E.g. for my ALC I used VelAdventure01 and the link to https://labs.geocaching.com/goto/VelAdventure01 is still working.

Just asking, is this identifier available in theory (even if it may not be usable)? Or the short link like `https://adventurelab.page.link/KSQT`? 

But probably it's better to continue in a dedicated issue...

_Originally posted by @fm-sys in https://github.com/cgeo/cgeo/issues/10372#issuecomment-820609621_

**Comment by baiti** [comment]
`https://adventurelab.page.link/KSQT`.....  the KSQT in this case, can only be used in the context of an URL to launch GC's app. It is absolutely useless for any other purpose.

There are two API calls that are relevant for us. The first just takes coordinates and comes back with a list of cache objects.

E.g.: `https://labs-api.geocaching.com/Api/Adventures/SearchV3?take=1&origin.longitude=48&origin.latitide=9`


The second call matters: `https://labs-api.geocaching.com/Api/Adventures/3b97bc3a-3581-42c2-a8f3-d9ababdae800`

I had started by putting the short thing above into `cache.geocode` and the second into `cache.uuid`.

And then I ran into the problems described here: https://github.com/cgeo/cgeo/pull/10372#issuecomment-820609621

The system expects: `cache.geocode = 3b97bc3a-3581-42c2-a8f3-d9ababdae800` as that value is needed everywhere where cache.geocode is referenced today and it is all good except that it is too long when it comes to rendeing it at the UI.

I would argue, that it does not make any sense to render it at all. No human can do anything with such a code.

**Comment by baiti** [comment]
Here is what we get from the search  API:

```
{
            "Id": "f4ea9d5f-5ded-4043-8ac3-65cd6b7e4752",
            "Title": "Swakopmund Adventure Lab",
            "KeyImageUrl": "https://gsmediadata.blob.core.windows.net/mediacontainer/25167c65-f4dc-4205-8e5a-e74731361e15",
            "SmartLink": "9521c103-faf2-4e9a-82e2-7a3a42667fc7",
            "DeepLink": "https://labs.geocaching.com/goto/9521c103-faf2-4e9a-82e2-7a3a42667fc7",
            "FirebaseDynamicLink": "https://adventurelab.page.link/f96j",
            "Description": "",
            "OwnerPublicGuid": "e22db7fa-6045-4d09-823c-0f2bda0d4f2e",
            "Visibility": 2,
            "CreatedUtc": "0001-01-01T00:00:00",
            "PublishedUtc": "2020-12-16T04:35:06.13",
            "IsArchived": false,
            "RatingsAverage": 4.7,
            "RatingsTotalCount": 7,
            "Location": {
                "Latitude": -22.67625,
                "Longitude": 14.5248,
                "Altitude": null
            },
            "StagesTotalCount": 5,
            "IsTest": false,
            "IsComplete": false
        }
```

`Id` is what I put into `cache.geocode`
the last part of `FirebaseDynamicLink` I put into `cache.uuid` and use the full link for "Open in Brower".

I believe that interface has changed recently. And there will be changes for sure. And they will happen without notice for sure.

I'd assume it doesn't matter for launching the app whether we use the FirebaseDynamicLink or the SmartLink. I'd have to test. I just used one that works and felt more comforable with the short identifier atin FirebaseDynamikLink

**Comment by Lineflyer** [comment]
As that "FIrebaseDynamicLink" links to the "lab detail page" its IMHO comparable to a GC-code.

Using that link suffix as geocode would IMHO be the nearest match to classical geocodes when it comes to usability, e.g. I can tell another cacher I recently made a nice adventure with code "K5QT" and he will be able to find it by using "adventurelab.page.link/" and appending the code. Similar to "coord.info/" and appending a GC-code.

It looks a bit as they even designed it as something comparable.

Regarding the possible sudden changes, which might occur we should make sure to use a fallback if that firebasethingy is missing.

**Comment by baiti** [comment]
@Lineflyer ... what do you expect to use this nice and short code for?

Please be aware, it is only useful for launching the App.

You cannot use this code to fetch cache details. Neither via the API nor can you scape any webpage using this code.

The cgeo application understands only one code, `geocode` and uses it at the UI and as a parameter to fetch detailed cache information.

For ALC there are now two code. One that is nice but useless, the other one looks ugly but is desperately needed and it is needed in the `cache.geocode` property, there is no way around it.

**Comment by baiti** [comment]
Do be precise: Using that suffix as geocode will not work. See here for a rationle: https://github.com/cgeo/cgeo/pull/10372#issuecomment-820609621

**Comment by Lineflyer** [comment]
> It is absolutely useless for any other purpose.
Mostly, but at least I can use https://adventurelab.page.link/KSQT in a plain web browser to view the adventures base info (which means...but thats another issue: Its another argument, that we could also show it despite any rules of any API...its available anyhow).

My main goal would be to have a more compact geocode, which can be used in the scenario I described in the duplicate issue:
User 1 says: There is a nice lab cache with code xyz
User 2: I will search for that geocode in the geocode search in my c:geo, then I can see where it is located.

**Comment by baiti** [comment]
> > It is absolutely useless for any other purpose.
> 
> Mostly, but at least I can use https://adventurelab.page.link/KSQT in a plain web browser to view the adventures base info (which means...but thats another issue: Its another argument, that we could also show it despite any rules of any API...its available anyhow).

Yes, however, the information that can be obtained by fetching data via this link is very sparse. You can get a title and a description. No publish date, no owner, not even header coordinates. Of course no information about stages, not even the number of stages,

> 
> My main goal however would be to have a more compact geocode, which can be used in the scenario I described in the duplicate issue:
> User 1 says: There is a nice lab cache with code xyz
> User 2: I will search for that geocode in the geocode search in my c:geo, then I can see where it is located.

Would you expect, that User 2 could punch in `code xyz` into cgeo somewhere and then make cgeo retrieve data related to that adventure from GC HQ's servers?

**Comment by Lineflyer** [comment]
> Would you expect, that User 2 could punch in code xyz into cgeo somewhere and then make cgeo retrieve data related to that adventure from GC HQ's servers?

Yes, same link for any classical geocode today:
Like I would say to you: yesterday I did a nice adventure. Its code is `XYZ`. Then you start your c:geo, go to search and enter `xyz` and you will get the adventure listing to work with it in c:geo (= store it to your ToDo, look up header coords on map, start Lab app).

**Comment by baiti** [comment]
> > Would you expect, that User 2 could punch in code xyz into cgeo somewhere and then make cgeo retrieve data related to that adventure from GC HQ's servers?
> 
> Yes, same link for any classical geocode today:
> Like I would say to you: yesterday I did a nice adventure. Its code is `XYZ`. Then you start your c:geo, go to search and enter `xyz` and you will get the adventure listing to work with it in c:geo (= store it to your ToDo, look up header coords on map, start Lab app).
> 

For GCxxxxx we do this each and every day and it only works becayse GCxxxxx can be used as an addressing component either in the scraping connector or in the partner API connector (should that ever come into existence),

XYZ on the otherhand has no meaning to GC HQ's servers. It is not possible to obtain any information about an Adventure using XYZ as a reference.

> As far as I tested, this seems to be already possible with the current long ALC geocode...its just unhandy.

Yes, that works, because that ugly code can directly be used as an addressing element for the public API.

**Comment by Lineflyer** [comment]
> Yes, that works, because that ugly code can directly be used as an addressing element for the public API. That long ID, as ugly as it is, is the semantic equivalent to the GC code as we know it from the traditional game

OK, this sounds as its not possible to shorten it and - at the same time - be able to expand it again into the original element needed towards the API..or is there any such reversible magic, which does not need individual lookup tables?

**Comment by baiti** [comment]
> > Yes, that works, because that ugly code can directly be used as an addressing element for the public API. That long ID, as ugly as it is, is the semantic equivalent to the GC code as we know it from the traditional game
> 
> OK, this sounds as its not possible to shorten it and - at the same time - be able to expand it again into the original element needed towards the API..or is there any such reversible magic, which does not need individual lookup tables?

It would only have a meaning if User 1 and User 2 had both previously fetched and persisted that Adventure on their devices. In that case we might introduce a new search capability into the overall search dialog.

**Comment by baiti** [comment]
> > Yes, that works, because that ugly code can directly be used as an addressing element for the public API. That long ID, as ugly as it is, is the semantic equivalent to the GC code as we know it from the traditional game
> 
> OK, this sounds as its not possible to shorten it and - at the same time - be able to expand it again into the original element needed towards the API..or is there any such reversible magic, which does not need individual lookup tables?
> 
> Edit:
> I really wonder why GS designed it that way without a proper reference code...

From their perspective it a perfectly proper reference code. The thing we know as "GC-code", an entity that human beings try to store in their head, is not required for the game called Adventure Lab. Nobody has a need to remember the UUID of an Adventure Lab cache vie their head. Nor punching it into any searchfield. Either they see a QR code, or a link they can click on or a dot on the map they can click on.

**Comment by baiti** [comment]
> > > Yes, that works, because that ugly code can directly be used as an addressing element for the public API. That long ID, as ugly as it is, is the semantic equivalent to the GC code as we know it from the traditional game
> > 
> > 
> > OK, this sounds as its not possible to shorten it and - at the same time - be able to expand it again into the original element needed towards the API..or is there any such reversible magic, which does not need individual lookup tables?
> 
> It would only have a meaning if User 1 and User 2 had both previously fetched and persisted that Adventure on their devices. In that case we might introduce a new search capability into the overall search dialog. ... and it would require that lookup table you have just mentioned while I was typing this ;)
> 
> We would not find a match in that table for any arbitrary adventure we hadn't seen and persisted before

The only reason why the current ALC implementation in cgeo renders that ugly ID on the UI is because it would require connector specific if then else logic in common code in order to elminiate that information form the UI if the cache type happens to be ALC

**Comment by baiti** [comment]
If you'd go to the discussions in the first PRs related to the subject you'd find that I started exactly where you started and I struggled a very long time, must have been weeks rather than days until I finally noticed it is just not worth it to try to maintain that association of a GC-code for ALCs

**Comment by Lineflyer** [comment]
SIde note:

> I'd assume it doesn't matter for launching the app whether we use the FirebaseDynamicLink or the SmartLink. I'd have to test. I just used one that works and felt more comforable with the short identifier atin FirebaseDynamikLink

I would tend to use the `FirebaseDynamicLink' for "open in browser" (as you already do as far as I understood) and use the `Smartlink` for launching the Adventure lab app.

**Comment by Lineflyer** [comment]
> If you'd go to the discussions in the first PRs related to the subject you'd find that I started exactly where you started and I struggled a very long time, must have been weeks rather than days until I finally noticed it is just not worth it to try to maintain that association of a GC-code for ALCs

Sounds indeed like that.
I suggest then to keep it as it is and leave this open and see how things (=public/partner API) evolve over time and see whether we ever get any better idea.

**Comment by baiti** [comment]
> SIde note:
> 
> > I'd assume it doesn't matter for launching the app whether we use the FirebaseDynamicLink or the SmartLink. I'd have to test. I just used one that works and felt more comforable with the short identifier atin FirebaseDynamikLink
> 
> I would tend to use the `FirebaseDynamicLink` for "open in browser" (as you already do as far as I understood) and use the `Smartlink` for launching the Adventure lab app.

There are three properties that have the term "link" in them:
```
 "SmartLinkString": null,
 "DeepLink": "https://labs.geocaching.com/goto/363c416e-7ca2-4911-9721-cd9cd319444a",
 "FirebaseDynamicLink": "https://adventurelab.page.link/11qf",
```
- `SmartLinkString` is sometimes `null` as you can see in the example above. I didn't even start to think about exploiting a resources that is not always there. If it is not `null`, it only holds a UUID string, not a complete link. I have not found any discriminator that would tell me under what conditions the `SmartLinkString` holds `null` or a valid id. Perhaps the descriminator is just the time when the adventure was published. Perhaps there was a time at which the adventure lab creator tooling decided to put something into that property and perhaps there was or is a time at which the adventure lab creator tooling decided not to put that information in because it is in fact redundant as it is the same informaiton that is already there in the `Id` property. After all, the API is public but not a supported API as we know and we also know that it may be changed at any time without notice.
 
- `DeepLink` opens a web page that has the QR code at the bottom. I have concluded that leading the user to that page via "Open in Browser" is useless as it does not provide a single bit of additional information over what is already available in cgeo's UI (except the QR code, which is useless from wihtin cgeo too). `DeepLink` sometimes points to a short identifier, "11qf" in the case above, sometimes a mnemonic string like: https://labs.geocaching.com/goto/VelAdventure01. It is not clear under which circumstances the code vs the mnemonic is used.

- `FirebaseDynamicLink` is the one rendered as a QR code on webpages. It is used to launch into the app.

To summarize from my perspective: The only target to launch into is "the app", regardless which link we use and regardless whether this is done by selecting "Open in browser" or via the "Open in Adventure Lab" at the bottom of the "Details" screen.

**Comment by Lineflyer** [comment]
Sounds all reasonable. So I guess your implementation is the best way.

**Comment by baiti** [comment]
> Sounds all reasonable. So I guess your implementation is the best way.

Yes I think so too, however, there is still an aspect of this thread we should continue to focus on: The geocode for ALC looks ugly and consumes additional real estate in list views as line wrapping opens an additional line.

The cleanest way of doing it would be `cache.getShortGeocode()`, which could be invoked independent from the cache type as it wouldn't do any change to cache types other than ALC. I had discussed this with @fm-sys at the top of the thread, however, it just didn't work out nicely as at the point where we would put this method invokation to, we'd need to put both, the full string and the truncated string and this would require interface changes, adding an additional parameter to an interface, maybe more than just one interface.

Another option would be to give up on the clean way of `cache.getShortGeocode()`, find the place where the data is bound to the UI and truncate the string there. I think this should work and I would have at least tried it, however, I have been unable so far to find the place in the code where this data binding is done. I may find it eventually but it would be much faster If one of the more experienced guys on the team pointed me to the right place. We could start with CacheDetailActivity first and then worry about list views later.

**Comment by Lineflyer** [comment]
> Another option would be to give up on the clean way of cache.getShortGeocode(), find the place where the data is bound to the UI and truncate the string there. I think this should work and I would have at least tried it, however, I have been unable so far to find the place in the code where this data binding is done. I may find it eventually but it would be much faster If one of the more experienced guys on the team pointed me to the right place. We could start with CacheDetailActivity first and then worry about list views later.

You mean something like truncating the Geocode at all places where it is shown on the UI to a certain amount of maximum characters? As all other geocodes are potentially short enough, it will only be used in the LC connector.

**Comment by baiti** [comment]
> You mean something like truncating the Geocode at all places where it is shown on the UI to a certain amount of maximum characters? As all other geocodes are potentially short enough, it will only be in action for caches of the LC connector.

Exactly that was the rationale. I had coded my short test for 10 characters. That doesn't hurt "normal" geocodes and makes it short enough to "work" in list views. It could perhaps even be memorized by the human brain and the existing "Search by geocode" even works today.

**Comment by fm-sys** [comment]
If we want to stay with the "ugly" geocode we could just use `singleline=true` and `end="ellipsis"` (or something like that). No need to progress the string programmly. But I haven't gave up showing the "correct" geocode yet :)

**Comment by baiti** [comment]
> If we want to stay with the "ugly" geocode we could just use `singleline=true` and `end="ellipsis"` (or something like that). No need to progress the string programmly. 

That would be cool. Where in the code would we make these changes? Or is that even somewhere in the XML? I'll go search .... this could even be a "quick fix" until we have something better.

> But I haven't gave up showing the "correct" geocode yet :)

Interesting ... have an idea yet?

**Comment by baiti** [comment]
Here are my findings so far: I have found the following 3 places where geocodes show up at the UI:

1. Click on the map, a popup shows up and the geocode is very ugly there and can potentially be shortened via 
```
android:maxLines="1"
android:ellipsize="end"
```
2.  CacheDetailActivity, here as ugly as it is, it fits on the screen in a single line and does not take away any real estate that would be used for other purposes if the code was shortened

3. In listviews. Here it hurts the most as it affects every list entry which uses now an additional line. However, here is also the problem. There is a single string that consists of `geocode + difficulty + terrain + cache type + premium`. The resulting string is bound to the UI element and the method shown above with maxLines and ellipsize would unfortunately truncate exactly the wrong parts in the string. So for this case, we programmatically need to shorten the geocode before the concatenation is done. I have not found the place in the code yet but will be searching for it.

**Comment by moving-bits** [comment]
@baiti 
IIRC this is `utils/Formatter.formatCacheInfoLong()`, you have a cache object there, so you can check for the connector responsible for it

**Comment by baiti** [comment]
> @baiti
> IIRC this is `utils/Formatter.formatCacheInfoLong()`, you have a cache object there, so you can check for the connector responsible for it

Perfect! .... That is the place where `cache.getShortGeocode()` will help ;) Will check in the afternoon but I am confident that this will address the ugliest case, #3 above.

**Comment by baiti** [comment]
I have a demo up and running that addresses all three cases mentioned above.

1. Click on the map - fits no on one line via the ellipsize annotation

![Screenshot from 2021-04-28 17-48-45](https://user-images.githubusercontent.com/5296333/116434000-501c2480-a84a-11eb-9693-cdf3ee56750e.png)

2. CacheDetail Activityy - change, just works

![Screenshot from 2021-04-28 17-49-11](https://user-images.githubusercontent.com/5296333/116434225-88236780-a84a-11eb-9ddb-767b397d8203.png)


3. List Views - shortened via cache.getShortGeocode()

![Screenshot from 2021-04-28 17-39-03](https://user-images.githubusercontent.com/5296333/116434439-b4d77f00-a84a-11eb-9e1a-46667d4c9196.png)

How about throwing it into a PR?

**Comment by Lineflyer** [comment]
Regarding 2.
One your device it fits, but not on all, therefore i guess it should be shortened as well.

**Comment by Lineflyer** [comment]
Regarding 1.
This notation has the disadvantage, that the popup will be maximized in its width. Couldn't you use the short Geocode there as well to keep this window from being unnecessary wide?

**Comment by baiti** [comment]
> Regarding 1.
> This notation has the disadvantage, that the popup will be maximized in its width. Couldn't you use the short Geocode there as well to keep this window from being unnecessary wide?

calling `cache.getShortGeocode()` is only possible if we have a reference to the `cache` object at the point where the data binding is done. Unfortunately that is not the case in some scenarios. Interface changes would be required such that an additional parameter (the shortened string) gets passed over these interfaces. Certainly not impossible. I could give it a try, however the required code changes I would consider a bit of a hack.

BTW: does it matter what the width of the popup is? Do we have an objective to render such a popup as slim as possible?

**Comment by baiti** [comment]
On my device (OnePlus 5T) the popup in case for a normal geocache renders a popup that is exactly the same size as the popup for an ALC cache.

Just to make sure that we're talking the same thing: It is the popup that opens upon pressing a cache icon on the map, right?

**Comment by Lineflyer** [comment]
> On my device (OnePlus 5T) for a normal geocache a popup is rendered that is exactly the same size as the popup for an ALC. That means, shortening the string, doesn't make the popup any smaller, neither in height nor in width

Just doublechecked it. You are right. It obviously limited to preserve a part of the underlying map left and right. So please disregard my comment.
